core/src/main/scala/com/databricks/labs/remorph/coverage/estimation/EstimationAnalyzer.scala

```diff
@@ -11,14 +14,159 @@ object SqlComplexity {
   case object COMPLEX extends SqlComplexity
   case object VERY_COMPLEX extends SqlComplexity
 
+  // TODO: Define the scores for each complexity level
+  def fromScore(score: Int): SqlComplexity = score match {
+    case s if s < 10 => LOW
+    case s if s < 20 => MEDIUM
+    case s if s < 30 => COMPLEX
+    case _ => VERY_COMPLEX
+  }
+
   implicit val rw: ReadWriter[SqlComplexity] = ReadWriter.merge(
     macroRW[SqlComplexity.LOW.type],
     macroRW[SqlComplexity.MEDIUM.type],
     macroRW[SqlComplexity.COMPLEX.type],
     macroRW[SqlComplexity.VERY_COMPLEX.type])
 }
 
+case class ComplexityEstimate(complexity: SqlComplexity, statementCount: Int, charCount: Int, lineCount: Int)
+case class SourceTextComplexity(lineCount: Int, textLength: Int)
+
+case class EstimationStatistics(
+    medianScore: Int,
+    meanScore: Double,
+    modeScore: Int,
+    stdDeviation: Double,
+    percentile25: Double,
+    percentile50: Double,
+    percentile75: Double,
+    geometricMeanScore: Double,
+    complexity: SqlComplexity)
+
+object EstimationStatistics {
+  implicit val rw: ReadWriter[EstimationStatistics] = macroRW
+}
+
 class EstimationAnalyzer extends LazyLogging {
   // TODO: We will do this and the rest of the analysis in the next PR
   def countStatements(plan: ir.LogicalPlan): Int = 1
+
+  def estimateComplexity(query: ExecutedQuery, plan: ir.LogicalPlan): ComplexityEstimate = {
+    val s = sourceTextComplexity(query.source)
+    ComplexityEstimate(SqlComplexity.LOW, countStatements(plan), s.textLength, s.lineCount)
+  }
+
+  def evaluateTree(node: TreeNode[_]): Int = {
+    evaluateTree(node, logicalPlanEvaluator, expressionEvaluator)
+  }
+
+  def evaluateTree(
```

Comment:
The various null-checks in this method look a bit "un-scala-ish". Did you actually stumble upon a case where `.children` returns `null`? If so, that's definitely a bug. A proper solution would be to define a default implementation for `TreeNode#children` that returns `Seq.empty`.

Reply:
I did not try and pin it down yet. I thought about what we might do here, but I think we want to report (and assign complexity to), malformed IR. Of course, we should be ensuring that anything that we do not support creates an UnsupportedXXXX and that perhaps we can ensure that in the Incomplete stuff in visitors - but that is complicated by sometimes visitChildren  being the correct thing to call instead of it being default implementation for an unimplemented visitor.